### PR TITLE
backend Pytorch support MIONetCartesianProd and PODMIONet

### PR DESCRIPTION
**Pull request by Jerry-Jzy**
## what is the purpose of this request
backend Pytorch support MIONetCartesianProd and PODMIONet

### CODE REVIEW

**Comment by lululxvi** [comment]
Reformat via black.

File: deepxde/nn/pytorch/mionet.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
+import torch
+from .nn import NN
+from .fnn import FNN
+from .. import activations
+from .. import initializers
+from .. import regularizers
+from ... import config
+from ...utils import timing
```
[the remark this answers is not in the document]
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/mionet.py]: fixed

File: deepxde/nn/pytorch/__init__.py [hunk 2]
```diff
@@ -1,7 +1,16 @@
...
-__all__ = ["DeepONetCartesianProd", "FNN", "NN", "PFNN", "PODDeepONet"]
+__all__ = [
+    "DeepONetCartesianProd",
+    "FNN",
+    "NN",
+    "PFNN",
+    "PODDeepONet",
+    "PODMIONet",
+    "MIONetCartesianProd",
+]
```
Comment (lululxvi) [inline deepxde/nn/pytorch/__init__.py:11]:
```
__all__ = [
    "DeepONetCartesianProd",
    "FNN",
    "MIONetCartesianProd",
    "NN",
    "PFNN",
    "PODDeepONet",
    "PODMIONet",
]
```

**Comment by lululxvi** [comment]
Have you tested MIONetCartesianProd and PODMIONet?

File: deepxde/nn/pytorch/mionet.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
+import torch
+
+from .nn import NN
+from .fnn import FNN
+from .. import activations
+
+
+class MIONetCartesianProd(NN):
+    """MIONet with two input functions and proper orthogonal decomposition (POD) for Cartesian product format."""
```
Comment (lululxvi) [inline deepxde/nn/pytorch/mionet.py:9]: Max length is 88.

**Comment by Jerry-Jzy** [comment]
> 

I have test MIONetCartesianProd

> Have you tested MIONetCartesianProd and PODMIONet?

yes

**Comment by Jerry-Jzy** [comment]
> 

PODMIONet has been tested

**Comment by lululxvi** [comment]
> PODMIONet has been tested

What is the accuracy of MIONet vs PODMIONet?

**Comment by Jerry-Jzy** [comment]
> > I have test MIONetCartesianProd
> 
> Can you obtain the same accuracy as Shuai's paper?

dont know why , i get better accuacy of MIONet

**Comment by Jerry-Jzy** [comment]
> > PODMIONet has been tested
> 
> What is the accuracy of MIONet vs PODMIONet?

POD has less accuracy

**Comment by lululxvi** [comment]
Details?

**Comment by Jerry-Jzy** [comment]
> 

not less, worse

File: deepxde/nn/pytorch/mionet.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
+import torch
+
+from .nn import NN
+from .fnn import FNN
+from .. import activations
```
Comment (lululxvi) [inline deepxde/nn/pytorch/mionet.py:5]:
```
from .fnn import FNN
from .nn import NN
from .. import activations
```

Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/mionet.py:5]: fixed

File: deepxde/nn/pytorch/mionet.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        x_func1 = inputs[0]
+        x_func2 = inputs[1]
+        x_loc = inputs[2]
+        # Branch net to encode the input function
+        y_func1 = self.branch1(x_func1)
+        y_func2 = self.branch2(x_func2)
+        # Trunk net to encode the domain of the output function
+        if self._input_transform is not None:
+            y_loc = self._input_transform(x_loc)
+        y_loc = self.activation_trunk(self.trunk(x_loc))
```
Comment (lululxvi) [inline deepxde/nn/pytorch/mionet.py:60]:
```python
y_loc = self.trunk(x_loc)
```

Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/mionet.py:60]: in tensorflow/deeponet.py , also use activation func after truck net
so I follow that way

File: deepxde/nn/pytorch/mionet.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        # Dot product
+        if y_func1.shape[-1] != y_func2.shape[-1]:
+            raise AssertionError(
+                "Output sizes of branch1 net and branch2 net do not match."
```
Comment (lululxvi) [inline deepxde/nn/pytorch/mionet.py:64]: branch1 net --> branch net1
branch2 net --> branch net2

File: deepxde/nn/pytorch/mionet.py (hunk 5 above)
Comment (lululxvi) [reply deepxde/nn/pytorch/mionet.py:60]: But this is MIONet.

File: deepxde/nn/pytorch/mionet.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+            )
+        y_func = torch.mul(y_func1, y_func2)
+        if self.trunk is None:
+            # POD only
+            y = torch.einsum("bi,ni->bn", y_func, self.pod_basis)
+        else:
+            y_loc = self.activation_trunk(self.trunk(x_loc))
```
Comment (lululxvi) [inline deepxde/nn/pytorch/mionet.py:145]: Also here.

Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/mionet.py]: fixed

File: deepxde/nn/pytorch/mionet.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+    def __init__(
+        self,
+        layer_sizes_branch1,
+        layer_sizes_branch2,
+        layer_sizes_trunk,
+        activation,
+        kernel_initializer,
+        regularization=None,
+        trunk_activation=False,
```
Comment (lululxvi) [inline deepxde/nn/pytorch/mionet.py:19]: Use trunk_last_activation

File: deepxde/nn/pytorch/mionet.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        # Branch net to encode the input function
+        y_func1 = self.branch1(x_func1)
+        y_func2 = self.branch2(x_func2)
+        # Trunk net to encode the domain of the output function
+        if self._input_transform is not None:
+            y_loc = self._input_transform(x_loc)
+        if self.trunk_activation:
+            y_loc = self.activation_trunk(self.trunk(x_loc))
+        else:
+            y_loc = self.trunk(x_loc)
```
Comment (lululxvi) [inline deepxde/nn/pytorch/mionet.py:65]:
```python
y_loc = self.trunk(x_loc)
if self.trunk_last_activation:
    y_loc = self.activation_trunk(y_loc)
```

File: deepxde/nn/pytorch/mionet.py (hunk 8 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/mionet.py]: fixed

File: deepxde/nn/pytorch/mionet.py (hunk 9 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/mionet.py]: fixed
